pkg/clustermesh/common/namespacewatcher/watcher.go

```diff
@@ -0,0 +1,507 @@
+	// IsGlobalNamespace returns true if the given namespace should be exported globally.
+	// Returns true by default for backwards compatibility if no specific configuration exists.
+	IsGlobalNamespace(namespace string) bool
+
+	// GetGlobalNamespaces returns the set of namespaces that are currently marked as global.
+	GetGlobalNamespaces() sets.Set[string]
+
+	// RegisterProcessor registers a processor that will be notified when namespace status changes
+	RegisterProcessor(processor NamespaceProcessor)
+
+	// IsFilteringActive returns true if namespace-based filtering is currently active
+	IsFilteringActive() bool
+
+	// IsGlobalService checks if a service is global considering both
+	// the service annotation and namespace filtering when active.
+	IsGlobalService(obj interface {
+		GetAnnotations() map[string]string
+		GetNamespace() string
+	}) bool
+
+	// IsSharedGlobalService checks if a service is both global and shared
+	IsSharedGlobalService(obj interface {
+		GetAnnotations() map[string]string
+		GetNamespace() string
+	}) bool
+}
+
+// NamespaceProcessor handles processing when namespace global status changes
+type NamespaceProcessor interface {
+	ProcessNamespaceChange(namespace string, isGlobal bool)
+	// GetAllNamespaces returns all namespaces that have resources in this processor's resource index
+	GetAllNamespaces() []string
+}
+
+const (
+	// OptClusterMeshDefaultGlobalNamespace is the name of the clustermesh-default-global-namespace option
+	OptClusterMeshDefaultGlobalNamespace = "clustermesh-default-global-namespace"
+)
+
+// Config configures the namespace watcher behavior.
+type Config struct {
+	// DefaultGlobalNamespace determines the default behavior for namespaces when
+	// namespace-based filtering is active (i.e., when at least one namespace is annotated).
+	// When true, namespaces are global by default unless annotated with clustermesh.cilium.io/global=false.
+	// When false, namespaces are local by default unless annotated with clustermesh.cilium.io/global=true.
+	DefaultGlobalNamespace bool `mapstructure:"clustermesh-default-global-namespace"`
+}
+
+// Flags implements cell.Flagger to register the clustermesh-default-global-namespace flag.
+func (cfg Config) Flags(flags *pflag.FlagSet) {
+	flags.Bool(OptClusterMeshDefaultGlobalNamespace, cfg.DefaultGlobalNamespace,
+		"Default behavior for namespaces when namespace-based filtering is active. "+
+			"When true, namespaces are global by default unless annotated with 'clustermesh.cilium.io/global=false'. "+
+			"When false, namespaces are local by default unless annotated with 'clustermesh.cilium.io/global=true'.")
+}
+
+type namespaceWatcher struct {
+	logger            *slog.Logger
+	config            Config
+	mu                lock.RWMutex
+	namespaceResource resource.Resource[*slim_corev1.Namespace]
+
+	// Processors for handling namespace changes
+	processors []NamespaceProcessor
+
+	// For testing: if true, use synchronous processing
+	syncProcessing bool
```

Comment:
why does testing require synchronous processing?

Reply:
Not required. Removed